src/python/pants/util/objects.py

```diff
@@ -7,15 +7,18 @@
 import sys
 from abc import abstractmethod
 from builtins import object, zip
-from collections import OrderedDict, namedtuple
+from collections import OrderedDict, deque, namedtuple
 
-from future.utils import PY2
+from future.utils import PY2, text_type
 from twitter.common.collections import OrderedSet
 
-from pants.util.memo import memoized, memoized_classproperty
+from pants.util.memo import memoized, memoized_classmethod, memoized_classproperty
 from pants.util.meta import AbstractClass
 
 
+# TODO: when we can restrict the python version to >= 3.6 in our python 3 shard, we can use the
```

Comment:
Sadly that's not a sufficient condition; we also need to have dropped py2 support entirely

Reply:
Ok, I was thinking there could be some parallel development (not too much) of py3-only code which would be using dataclasses. I will make the comment more clear.

Reply:
Parallel development of Py2 and Py3 branches is heavily discouraged in Python community, so might not be the best approach. It used to be the recommend approach in the early days, and it led to extreme complexity for library developers.

So, the new best practice is using Python 3 first code that still works on Python 2, which is what we've been doing with the `future` library and backports.